File: 046 Permutations/permutations.cs
public class Solution
{	// Author: Xinyu Chen
    public IList<IList<int>> Permute(int[] nums)
    {
        var result = new List<IList<int>>();
        DoPermute(new List<int>(nums), 0, result);
        return result;
    }
    public void DoPermute(IList<int> nums, int index, IList<IList<int>> result)
    {
        if (index == nums.Count)
        {
            result.Add(new List<int>(nums));
            return;
        }
        for (int i = index; i < nums.Count; ++i)
        {
            int temp = nums[i];
            nums[i] = nums[index];
            nums[index] = temp;
            DoPermute(nums, index + 1, result);
            nums[index] = nums[i];
            nums[i] = temp;
        }
    }
}

File: 046 Permutations/permutations.cpp
class Solution {  // Author: Xinyu Chen
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> group;
        dfs(nums, group, result);
        return result;
    }
    
    void dfs(vector<int>& nums, vector<int>& group, vector<vector<int>>& result) {
        if (group.size() == nums.size()) {
            result.push_back(group);
            return;
        }
        for (auto n : nums) {
            if (find(group.begin(), group.end(), n) == group.end()) {
                group.push_back(n);
                dfs(nums, group, result);
                group.pop_back();
            }
        }
    }
};
